rclpy/rclpy/node.py

```diff
@@ -682,25 +682,180 @@ def _set_parameters_atomically(
+
+        This method sets the type in the descriptor to match the parameter type.
+        If a descriptor is provided, its name will be set to the name of the parameter.
+
+        :param parameter: Parameter to be checked.
+        :param descriptor: Descriptor to apply. If None, the stored descriptor for the given
+            parameter's name is used instead.
+        :param check_read_only: True if read-only check has to be applied.
+        :return: SetParametersResult; successful if checks passed, unsuccessful otherwise.
+        :raises: ParameterNotDeclaredException if a descriptor is not provided, the given parameter
+            name had not been declared and undeclared parameters are not allowed.
+        """
         if descriptor is None:
-            descriptor = self._descriptors[parameter.name]
+            descriptor = self.describe_parameter(parameter.name)
+        else:
+            descriptor.name = parameter.name
+
+        # The type in the descriptor has to match the type of the parameter.
+        descriptor.type = parameter.type_.value
+
+        if check_read_only and descriptor.read_only:
+            return SetParametersResult(
+                successful=False,
+                reason='Trying to set a read-only parameter: {}.'.format(parameter.name))
+
+        if parameter.type_ == Parameter.Type.INTEGER and descriptor.integer_range:
+            return self._apply_integer_range(parameter, descriptor.integer_range[0])
+
+        if parameter.type_ == Parameter.Type.DOUBLE and descriptor.floating_point_range:
+            return self._apply_floating_point_range(parameter, descriptor.floating_point_range[0])
+
+        return SetParametersResult(successful=True)
+
+    def _apply_integer_range(
+        self,
+        parameter: Parameter,
+        integer_range: IntegerRange
+    ) -> SetParametersResult:
+        min_value = min(integer_range.from_value, integer_range.to_value)
+        max_value = max(integer_range.from_value, integer_range.to_value)
+
+        # Values in the edge are always OK.
+        if parameter.value == min_value or parameter.value == max_value:
+            return SetParametersResult(successful=True)
 
-        self._parameters[parameter.name] = parameter
-        return True
+        if parameter.value < min_value or parameter.value > max_value:
+            return SetParametersResult(
+                successful=False,
+                reason='Parameter {} out of range\n'
+                       'Min: {}, Max: {}, value: {}'.format(
+                            parameter.name, min_value, max_value, parameter.value
+                        )
+            )
+
+        if integer_range.step != 0 and (parameter.value - min_value) % integer_range.step != 0:
+            return SetParametersResult(
+                successful=False,
+                reason='The distance between the parameter value for {}'
+                       'is not a multiple of step.\n'
```

Comment:
The distance between the parameter value, and what ...?? It sounds strange to me.
Maybe, `the parameter value is not one of the allowed steps`.

Reply:
Thanks; addressed in https://github.com/ros2/rclpy/pull/357/commits/7381f62aeade16395c33dee943d7a5aa94173062.